rust/chains/abacus-ethereum/src/provider.rs

```diff
@@ -0,0 +1,119 @@
+use async_trait::async_trait;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use ethers::prelude::{Middleware, H256, U256};
+use eyre::eyre;
+
+use crate::MakeableWithProvider;
+use abacus_core::{
+    AbacusChain, AbacusProvider, BlockInfo, ContractLocator, TxnInfo, TxnReceiptInfo,
+};
+
+/// Connection to an ethereum provider. Useful for querying information about
+/// the blockchain.
+#[derive(Debug, Clone)]
+pub struct EthereumProvider<M>
+where
+    M: Middleware,
+{
+    provider: Arc<M>,
+    chain_name: String,
+    domain: u32,
+}
+
+impl<M> AbacusChain for EthereumProvider<M>
+where
+    M: Middleware + 'static,
+{
+    fn chain_name(&self) -> &str {
+        &self.chain_name
+    }
+
+    fn local_domain(&self) -> u32 {
+        self.domain
+    }
+}
+
+#[async_trait]
+impl<M> AbacusProvider for EthereumProvider<M>
+where
+    M: Middleware + 'static,
+{
+    async fn get_block_by_hash(&self, hash: &H256) -> eyre::Result<BlockInfo> {
+        let block = self
+            .provider
+            .get_block(*hash)
+            .await?
+            .ok_or_else(|| eyre!("Could not find block with hash {}", hash))?;
+        assert_eq!(block.hash.as_ref().unwrap(), hash);
+        debug_assert_eq!(U256::from(block.timestamp.as_u64()), block.timestamp);
```

Comment:
any need for this? I think it's safe to assume that the unix time in seconds won't ever overflow a u64

Reply:
This is a `debug_assert` which means that it is not in the release version. While I do agree this should never happen, I am trying to be in the habit of verifying my assumptions which _has_ already caught a few issues. Not this one specifically but it seems useful.

As an aside, WTF do they think it needs to be a u256?

Reply:
ah okay, thanks for clarifying that it won't be in the release version, happy to keep as-is